stan/math/rev/fun/dot_product.hpp

```diff
@@ -31,40 +31,44 @@ namespace math {
  * @return Dot product of the vectors.
  * @throw std::domain_error if sizes of v1 and v2 do not match.
  */
-template <typename T1, typename T2, require_all_container_t<T1, T2>* = nullptr,
-          require_any_vt_var<T1, T2>* = nullptr>
+template <typename T1, typename T2, require_all_vector_t<T1, T2>* = nullptr,
+          require_all_not_std_vector_t<T1, T2>* = nullptr,
+          require_any_st_var<T1, T2>* = nullptr>
 inline var dot_product(const T1& v1, const T2& v2) {
   check_matching_sizes("dot_product", "v1", v1, "v2", v2);
+
+  if (v1.size() == 0) {
+    return 0.0;
+  }
+
   if (!is_constant<T1>::value && !is_constant<T2>::value) {
-    arena_t<vector_v> v1_arena = as_column_vector_or_scalar(v1);
-    arena_t<vector_v> v2_arena = as_column_vector_or_scalar(v2);
-    var res(v1_arena.val().dot(v2_arena.val()));
-    reverse_pass_callback([v1_arena, v2_arena, res]() mutable {
-      for (Eigen::Index i = 0; i < v1_arena.size(); ++i) {
-        const auto res_adj = res.adj();
-        v1_arena.coeffRef(i).adj() += res_adj * v2_arena.coeffRef(i).val();
-        v2_arena.coeffRef(i).adj() += res_adj * v1_arena.coeffRef(i).val();
-      }
-    });
-    return res;
+    arena_t<promote_scalar_t<var, T1>> v1_arena = v1;
+    arena_t<promote_scalar_t<var, T2>> v2_arena = v2;
+    return make_callback_var(
```

Comment:
I think I should be doing:

```
eigen_results(v1_arena.adj().array(), v2_arena.adj().array())
  += eigen_expressions(v2_arena.val().array() * res_adj,
                       v1_arena.val().array() * res_adj);
```

but it's segfaulting at runtime.

Reply:
These arrays can come from mixes of column vectors/row vectors btw. I assume that's somehow related.

Reply:
> I think I should be doing

Well you can't, because it is getting reverted.

Reply:
RIP